### ISSUE TITLE
How to use LM algorithm in a custom train loop with custom loss function?

### ISSUE THREAD

**Issue by mrleiyz** [fabiodimarco/tf-levenberg-marquardt#16]
Dear Sir or Madam,
  Thanks for your endeavor to develop such code.
  And, I try to use it in a custom training loop, and the loss function is also custom (except for y_true, and y_pred, it has other input variables).   So, what should I do to solve this problem by modifying your code?

 Hope for your response ASAP.
 Thanks a lot . @fabiodimarco

**Comment by fabiodimarco** [comment]
Hi,

to define a custom training loop you can create an instance of the `Trainer` class defined al line 273. Inside your training loop you can then call the member function `train_step` defined at line 593.

To use a custom loss function you can give a look at how I defined some of them. As an example you can look at: `MeanSquaredError` line 28 or `CategoricalCrossentropy` line 80. There are more in the same file that you can look at.
So the new loss function you define needs to have a `call` method that implements your loss and a `residual` which is used inside the LM optimiser.

The residual has to be defined so that:
loss = mean(residual^2)

It does not need to be literally like that, you can implement a more stable and computationally efficient expression as long as the final results is the one above.

Let me know if it helps.

**Comment by fabiodimarco** [comment]
Since w1 is a fixed constant you do not need to pass it as a parameter of the function, but you can save it as a member variable of your custom loss class.

```
class CustomLoss(tf.keras.losses.Loss):
    def __init__(self,
                 w1,
                 reduction=tf.keras.losses.Reduction.AUTO,
                 name='custom_loss'):
        super(CustomLoss, self).__init__(
            reduction=reduction,
            name=name)
       self.w1 = w1

    def call(self, y_true, y_pred):
        y_pred = tf.convert_to_tensor(y_pred)
        y_true = tf.cast(y_true, y_pred.dtype)

        loss1 = mse(y_pred , y_true)
        loss2 = custom_loss(y_pred , self.w1)
        loss_tol = loss1 + loss2

       return loss_tol

    def residuals(self, y_true, y_pred):
        y_pred = tf.convert_to_tensor(y_pred)
        y_true = tf.cast(y_true, y_pred.dtype)
        
       # you have to write here the code according to how you custom loss is defined
       # so that: loss = mean(residuals^2)
```

**Comment by mrleiyz** [comment]
Thank you for your patience indication!! It is very helpful!

I will try to realize it.

**Comment by fabiodimarco** [comment]
No problem. Let me know how it goes.

**Comment by mrleiyz** [comment]
Hi, Fabio, 
Following your suggestions, today I try to realize it, I have mastered how to define the custom loss function and use the Class Trainer, but I find a new problem. Yesterday, in my problem statements, I forgot there is an input variable in my custom loss function, namely, the custom loss not only relies on the Y_true, Y_pred, and some fixed constants but also needs the input variable X_train. 

So, I think the method of  model.compile / model_wrapper.compile cannot achieve my goal, because such method seems to determine the loss before the model training, but my goal is to introduce X_train to calculate the custom loss during the training process, meanwhile, the loss using in compile method inherits from tf.keras.losses which only accept output variables (Y_true and Y_pred).

What about your advice? 
Hope for your response ASAP.
Thanks a lot !!~~

**Comment by fabiodimarco** [comment]
A simple workaround is to include `X_train` in your `Y` data. So that `Y_true` is something like `tf.stack([X_train, Y_train], axis=-1)` or similar. I do not know the dimensionality of your data.

**Comment by mrleiyz** [comment]
Uhmm..., in my task, the data dimensions of X_train, and Y_train are not the same,  and at first, I need to normalize the input and label, then during the training process, namely the calculation of custom, I need to inverse-normalize the input, label and predicted label. So the operation could be complex.  Maybe I need to study the ModelWrapper carefully.

 Anyway,  thanks for your advice,  I will try it, I hope I can bring good news to you.

**Comment by fabiodimarco** [comment]
If X and Y have a different dimension then you can have Y_train to be a list or a tuple of tensors (X, Y).

In order to use the ModelWrapper, I would consider trying to place all the extra operation that you need to do during training inside the CustomLoss or inside the Model itself.
In that way you can use all the callback provided by keras to save chekpoints, logging etc.

**Comment by mrleiyz** [comment]
Hi, Fabio,  I am trying to realize the code. By the way, there are two questions I want to discuss with you.
The first one is, except to include X_train in Y data (as you mentioned above), how to pass other variables to model.fit ?  it seems only accept two parameters, X and Y.

The second one is,  I know the LM algorithm is used in the neural network toolbox of MATLAB in the early stage, in those years, we don't develop the technique of deep learning or deep neural network. Nowadays, we usually adopt deep networks, by google search, I know some people say the LM algorithm is useful for the shallow network with few neurons, but for the deep network with numerous trainable parameters, the algorithm has poor performance. I also see your test example is a simple nonlinear fitting, so, are you testing this algorithm with a more complex situation, such as image recognition or natural language process?

I also find an issue about using the algorithm in PINN, which is an interesting topic, but some physical problems or engineering problems could need a deep network, for such a situation, whether LM algorithm can get a better result than gradient descent?

Thanks.

**Comment by fabiodimarco** [comment]
1) No `model.fit` only accept X and Y, but since your loss is custom you can include into Y all the data you want even a dictionary so that you can access all the data you need to compute the loss (that is the way to do it).

2) The LM algorithm is mainly used to train shallow network as they have a small number of parameters.
The problem with LM is that the computational complexity is N^3 with respect to the number of parameters or the batch size depending on which one of the two is larger. You can find the details about it in the **Memory, Speed and Convergence considerations** section in the readme of the my repo.
You can tradeoff the complexity of LM by reducing the batch size, but the more you reduce it the more you are going towards a simple stochastic gradient descent and hence losing the advantages of LM.
So even if LM has better **per step** convergence rate (and sometimes can also converge to better training losses), the computational speed of computing each step for large models is the main reason gradient descent is still the standard for training large neural networks.
